text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,238 @@
+- Feature Name: extern-existential-type
+- Start Date: 2018-6-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+An extension of [#2071](https://github.com/rust-lang/rfcs/pull/2071)'s `existential type` where the definition can live in a different crate than the declaration, rather than the same module.
+This is a crucial tool untangling for untangling dependencies within `std` and other libraries at the root of the ecosystem concerning global resources.
+
+# Motivation
+[motivation]: #motivation
+
+We have a number of situations where one crate defines an interface, and a different crate implements that interface with a canonical singleton:
+
+ - [`core::alloc::GlobalAlloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html), chosen with [`#[global_allocator]`](https://doc.rust-lang.org/1.23.0/unstable-book/language-features/global-allocator.html)
+ - `panic_fmt` chosen with [`#[panic_implementation]`](https://github.com/rust-lang/rfcs/blob/master/text/2070-panic-implementation.md)
+ - The OOM hook, modified with [`std::alloc::{set,take}_alloc_error_hook`](https://doc.rust-lang.org/nightly/std/alloc/fn.set_alloc_error_hook.html)
+ - [`std::collections::hash_map::RandomState`](https://doc.rust-lang.org/std/collections/hash_map/struct.RandomState.html), if https://github.com/rust-lang/rust/pull/51846 is merged, the `hashmap_random_keys` lang item
+ - [`log::Log`](https://docs.rs/log/0.4.3/log/trait.Log.html) set with [`log::set_logger`](https://docs.rs/log/0.4.3/log/fn.set_logger.html)
```

Comment:
Logging can require resources such as files, network connections, run-time configuration data, etc. so it's difficult to make some logger implementations into `static`s. In theory everything could be initalized lazily on first logging call (though then you haven't eliminated the overhead of dynamic dispatch!), but this mostly just shifting the problem (and run-time costs) around and in more complex scenarios -- e.g. when you want to read a configuration file to determine what *kind* of logging to do -- this can require making much more state global than currently necessary. It also affects all other control flow surrounding logger initialization, e.g. error reporting from being unable to open a file for logging now has to be moved into the lazy-initialization code.

Reply:
Not every case needs initialization. And example is serial port logging for embedded.

For the logging case there's these options:

1. Just as local allocation exists, there should be `*_in` variants of the macros that allow passing around a local logger. One should never be forced into using a singleton.

2. The case where one really wants a single *xxx*, such that a static is nice because any passed pointer / value is just overhead, but also wants to be sure the *xxx*  is initialized first, is heavily explored by @japaric and the rest of the working group. In general, something still needs to be passed around, but it can just be a ZST "token" indicating initialization is complete. So this is just a riff off the above. My stuff still helps if you want to be monophonic over the token type.

3. Recreate today with a mutable singleton and manually initialize it. Unlike today this is opt-in. Something like `lazy_static` + my stuff can make this more ergonomic.

4. Lazy init as you mention. Actually `lazy_static` crate + my stuff makes this decently ergonomic.

So overall I think mine is still value for not needing to cfg around the no-std case my offering a better separation of concerns between functionality and its "singletonness", while inciting one to make the singleton aspect optional.

Reply:
> Not every case needs initialization. And example is serial port logging for embedded.

Yes, obviously.

> Just as local allocation exists, there should be *_in variants of the macros that allow passing around a local logger. One should never be forced into using a singleton.

While in theory I commend getting rid of global state, the fact is that such an API for logging is entirely hypothetical while the "global logger" API is pervasive in the ecosystem. If it's a huge pain to replace the current global singleton with a different mechanism for installing the same global singleton, everyone who (for whatever reasons) uses the global singleton will rightly prefer the current way of installing it over what this RFC presents.

Besides, this argument cuts both ways: all the code that avoids global singletons doesn't need this RFC either.

> Recreate today with a mutable singleton and manually initialize it. Unlike today this is opt-in. Something like lazy_static + my stuff can make this more ergonomic.

As with the OOM hook, the question is how commonly this would be done. If it's quite common, the benefit of external existential traits for this use case is diminished.

> Lazy init as you mention. Actually lazy_static crate + my stuff makes this decently ergonomic.

lazy_init is nice but it doesn't help you get the data needed for initialization (e.g., user configuration) into the block doing the initialization, as it can only reference other globals, not e.g. a local binding created in `main`. That's what I meant by requiring more state to be made global.

---

None of this is a fatal objection to using external existential types for logging. I am just saying that for some quite common use cases, it won't have all the claimed benefits and indeed some downsides as well.

Reply:
What downsides? The *worst case* of manual initialization forced by effects and user config works exactly like today.

Reply:
The downside is that any logger impementations that needs it has to separately go through the work of (and accept the risk of bugs in) emulating today's behavior. (Plus the churn of changing APIs.)

Reply:
@rkruppe liblog can come with a `ManuallyInitializedLogger<L>` that does all that for you.

Reply:
Hm, true. It's still not clear to me that a significant portion of liblog users would do anything different, but let's leave that discussion at https://github.com/rust-lang/rfcs/pull/2492#discussion_r199555213